Building TomoPy from source on Windows

I'm trying to build TomoPy from source on a fresh python 3.7 install (from python.org) in a Win10 environment. Running into some build issues and wondering if anybody can give me some tips? Here is where I am now:

1. I have installed MinGW and placed the binaries in my PATH following this procedure:

[https://www.hellocodies.com/how-to-install-mingw-compiler-on-windows/](https://www.hellocodies.com/how-to-install-mingw-compiler-on-windows/
)

2. I copied in `make` following the suggestion here:
[https://stackoverflow.com/questions/32127524/how-to-install-and-use-make-in-windows](https://stackoverflow.com/questions/32127524/how-to-install-and-use-make-in-windows)

3. Going through the win-37.yml file, I was able to `pip install`:
- h5py
- mkl
- mkl-devel
- nose
- numexpr
- numpy
- libopencv
- pywavelets
- scikit-build
- scikit-image
- scipy
- setup_tools_scm
- setup_tools_scm_git_archive
- six

4. Since `dxchange` is not available via PyPl, I cloned its repo and successfully built/installed from source. Git... that... dirt off your shoulder...

-- So far, so good! But now things start to go awry... --

5. The package `mkl_fft` does not seem to be available for python 3.7 via PyPl and doesn't seem to be supported beyond 3.6 at the moment:

https://stackoverflow.com/questions/59854306/python3-7-missing-pip-package

Trying to build this one from source throws errors, so right now mkl_fft is not installed. Still a WIP... the errors here are a topic for a different channel I suppose, but if there's a non-conda shortcut for mkl_fft I am all ears. 

6. The dependency `vs2017_win-64` appears to be an Anaconda package. I do have Visual Studio 2017 installed that I'm using for some other dev projects... what does tomopy need from there and can I serve it up via environment variable instead?

At the moment, when I run `pip install .` in the cloned tomopy directory the build fails with the attached output. The errors seem to be thrown by `cmake` though I also see failures here:
```
    -- Performing Test c_fp_model_precise - Failed
    -- Performing Test cxx_fp_model_precise - Failed
```

Full build output is here:
[tomopy-build-output.txt](https://github.com/tomopy/tomopy/files/4737479/tomopy-build-output.txt)

Is this purely an issue of missing dependencies (see 5 & 6 above), or is there a configuration step I'm missing here?

In the meantime, I'll try making a parallel python 3.6 installation to see if that helps resolve some missing dependencies (and, consequently, facilitate the tomopy build).

I have now installed CMake and added it to my PATH.  When I try to build tomopy now this is the error I get:
```
(tomopy) J:\epics\tomography\tomopy>python setup.py develop


--------------------------------------------------------------------------------
-- Trying "Ninja (Visual Studio 15 2017 Win64 v141)" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
-- Configuring incomplete, errors occurred!
See also "J:/epics/tomography/tomopy/_cmake_test_compile/build/CMakeFiles/CMakeOutput.log".
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Ninja (Visual Studio 15 2017 Win64 v141)" generator - failure
--------------------------------------------------------------------------------



--------------------------------------------------------------------------------
-- Trying "Visual Studio 15 2017 Win64 v141" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- The C compiler identification is MSVC 19.13.26129.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.13.26128/bin/Hostx86/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.13.26128/bin/Hostx86/x64/cl.exe - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: J:/epics/tomography/tomopy/_cmake_test_compile/build
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Visual Studio 15 2017 Win64 v141" generator - success
--------------------------------------------------------------------------------

Configuring Project
  Working directory:
    J:\epics\tomography\tomopy\_skbuild\win-amd64-3.6\cmake-build
  Command:
    cmake 'J:\epics\tomography\tomopy' -G 'Visual Studio 15 2017 Win64' '-DCMAKE_INSTALL_PREFIX:PATH=J:\epics\tomography\tomopy\_skbuild\win-amd64-3.6\cmake-install' '-DPYTHON_EXECUTABLE:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\python.exe' -DPYTHON_VERSION_STRING:STRING=3.6.10 '-DPYTHON_INCLUDE_DIR:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\Include' '-DPYTHON_LIBRARY:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\libs\python36.lib' -DSKBUILD:BOOL=TRUE '-DCMAKE_MODULE_PATH:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\lib\site-packages\skbuild\resources\cmake' -DCMAKE_BUILD_TYPE:STRING=Release

Using scikit-build. cmake: C:/Program Files/CMake/bin/cmake.exe...
-- The C compiler identification is MSVC 19.13.26129.0
-- The CXX compiler identification is MSVC 19.13.26129.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.13.26128/bin/Hostx86/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.13.26128/bin/Hostx86/x64/cl.exe - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.13.26128/bin/Hostx86/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.13.26128/bin/Hostx86/x64/cl.exe - broken
CMake Error at C:/Program Files/CMake/share/cmake-3.17/Modules/CMakeTestCXXCompiler.cmake:53 (message):
  The C++ compiler

    "C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.13.26128/bin/Hostx86/x64/cl.exe"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: J:/epics/tomography/tomopy/_skbuild/win-amd64-3.6/cmake-build/CMakeFiles/CMakeTmp

    Run Build Command(s):C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/MSBuild/15.0/Bin/MSBuild.exe cmTC_d7474.vcxproj /p:Configuration=Debug /p:Platform=x64 /p:VisualStudioVersion=15.0 /v:m && Microsoft (R) Build Engine version 15.6.85.37198 for .NET Framework
    Copyright (C) Microsoft Corporation. All rights reserved.

      Microsoft (R) C/C++ Optimizing Compiler Version 19.13.26129 for x64
      Copyright (C) Microsoft Corporation.  All rights reserved.

      cl /c /Zi /W3 /WX- /diagnostics:classic /Od /Ob0 /D WIN32 /D _WINDOWS /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"cmTC_d7474.dir\Debug\\" /Fd"cmTC_d7474.dir\Debug\vc141.pdb" /Gd /TP /errorReport:queue "J:\epics\tomography\tomopy\_skbuild\win-amd64-3.6\cmake-build\CMakeFiles\CMakeTmp\testCXXCompiler.cxx"
      testCXXCompiler.cxx

    LINK : fatal error LNK1104: cannot open file 'J:\epics\tomography\tomopy\_skbuild\win-amd64-3.6\cmake-build\CMakeFiles\CMakeTmp\Debug\cmTC_d7474.exe' [J:\epics\tomography\tomopy\_skbuild\win-amd64-3.6\cmake-build\CMakeFiles\CMakeTmp\cmTC_d7474.vcxproj]





  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:27 (project)


-- Configuring incomplete, errors occurred!
See also "J:/epics/tomography/tomopy/_skbuild/win-amd64-3.6/cmake-build/CMakeFiles/CMakeOutput.log".
See also "J:/epics/tomography/tomopy/_skbuild/win-amd64-3.6/cmake-build/CMakeFiles/CMakeError.log".
Traceback (most recent call last):
  File "C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\lib\site-packages\skbuild\setuptools_wrap.py", line 586, in setup
    languages=cmake_languages
  File "C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\lib\site-packages\skbuild\cmaker.py", line 240, in configure
    os.path.abspath(CMAKE_BUILD_DIR())))

An error occurred while configuring with CMake.
  Command:
    cmake 'J:\epics\tomography\tomopy' -G 'Visual Studio 15 2017 Win64' '-DCMAKE_INSTALL_PREFIX:PATH=J:\epics\tomography\tomopy\_skbuild\win-amd64-3.6\cmake-install' '-DPYTHON_EXECUTABLE:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\python.exe' -DPYTHON_VERSION_STRING:STRING=3.6.10 '-DPYTHON_INCLUDE_DIR:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\Include' '-DPYTHON_LIBRARY:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\libs\python36.lib' -DSKBUILD:BOOL=TRUE '-DCMAKE_MODULE_PATH:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\lib\site-packages\skbuild\resources\cmake' -DCMAKE_BUILD_TYPE:STRING=Release
  Source directory:
    J:\epics\tomography\tomopy
  Working directory:
    J:\epics\tomography\tomopy\_skbuild\win-amd64-3.6\cmake-build
Please see CMake's output for more information.
```

It looks like it fails when I tries to compile a test program: 
```
J:\epics\tomography\tomopy\_skbuild\win-amd64-3.6\cmake-build\CMakeFiles\CMakeTmp\testCXXCompiler.cxx
```
But that test program does not seem to exist in the directory it is building in:
```
 Directory of J:\epics\tomography\tomopy\_skbuild\win-amd64-3.6\cmake-build\CMakeFiles\CMakeTmp

06/22/2020  02:21 PM    <DIR>          .
06/22/2020  02:21 PM    <DIR>          ..
               0 File(s)              0 bytes
               2 Dir(s)  219,413,360,640 bytes free
```
So that CMakeTmp directory is empty, it does not contain the test file.

Any idea why not?

> The prefix error is to prevent you from overwriting and existing environment. It occurred because you already have an environment named tomopy. You can either remove the previous environment or specify a different name.

Thanks for the clarification.  I copied  win-36.yml to win-36-devel.yml and changed the environment name to tomopy_devel.  I created and activated that environment. Now I see cmake in the conda list.
```
(tomopy_devel) J:\epics\tomography\tomopy>conda list
# packages in environment at C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel:
#
# Name                    Version                   Build  Channel
blas                      1.0                         mkl
blosc                     1.19.0               ha925a31_0    conda-forge
brotli                    1.0.7             h33f27b4_1002    conda-forge
brotlipy                  0.7.0           py36h779f372_1000    conda-forge
bzip2                     1.0.8                hfa6e2cd_2    conda-forge
ca-certificates           2020.6.20            hecda079_0    conda-forge
certifi                   2020.6.20        py36h9f0ad1d_0    conda-forge
cffi                      1.14.0           py36ha419a9e_0    conda-forge
cftime                    1.1.3            py36h7725771_0    conda-forge
chardet                   3.0.4           py36h9f0ad1d_1006    conda-forge
charls                    2.1.0                h33f27b4_2    conda-forge
cloudpickle               1.4.1                      py_0    conda-forge
cmake                     3.17.0               he025d50_0    conda-forge
coverage                  5.1              py36h779f372_0    conda-forge
cryptography              2.9.2            py36hef61171_0    conda-forge
curl                      7.69.1               h1dcc11c_0    conda-forge
cycler                    0.10.0                     py_2    conda-forge
cytoolz                   0.10.1           py36hfa6e2cd_0    conda-forge
dask-core                 2.19.0                     py_0    conda-forge
decorator                 4.4.2                      py_0    conda-forge
distro                    1.5.0              pyh9f0ad1d_0    conda-forge
dxchange                  0.1.5                   py_1001    conda-forge
dxfile                    0.5                        py_1    conda-forge
edffile                   5.0.0           pyh24bf2e0_1000    conda-forge
freetype                  2.10.2               hd328e21_0    conda-forge
giflib                    5.2.1                h2fa13f4_2    conda-forge
git                       2.27.0                        0    conda-forge
h5py                      2.10.0          nompi_py36h422b98e_102    conda-forge
hdf4                      4.2.13            hf8e6fe8_1003    conda-forge
hdf5                      1.10.5          nompi_ha405e13_1104    conda-forge
icc_rt                    2019.0.0             h0cc432a_1
icu                       64.2                 he025d50_1    conda-forge
idna                      2.9                        py_1    conda-forge
imagecodecs               2020.5.30        py36h759e2a0_1    conda-forge
imageio                   2.8.0                      py_0    conda-forge
intel-openmp              2019.4                      245
jpeg                      9d                   he774522_0    conda-forge
jxrlib                    1.1                  hfa6e2cd_2    conda-forge
kiwisolver                1.2.0            py36h246c5b5_0    conda-forge
krb5                      1.17.1               hdd46e55_0    conda-forge
lcms2                     2.11                 he1115b7_0    conda-forge
libaec                    1.0.4                he025d50_1    conda-forge
libblas                   3.8.0                    14_mkl    conda-forge
libcblas                  3.8.0                    14_mkl    conda-forge
libclang                  9.0.1           default_hf44288c_0    conda-forge
libcurl                   7.69.1               h1dcc11c_0    conda-forge
liblapack                 3.8.0                    14_mkl    conda-forge
liblapacke                3.8.0                    14_mkl    conda-forge
libnetcdf                 4.7.4           nompi_hc957ea6_101    conda-forge
libopencv                 4.3.0                    py36_1    conda-forge
libpng                    1.6.37               hfe6a214_1    conda-forge
libssh2                   1.9.0                h3235a2c_2    conda-forge
libtiff                   4.1.0                h885aae3_6    conda-forge
libwebp-base              1.1.0                hfa6e2cd_3    conda-forge
libzopfli                 1.0.3                ha925a31_0    conda-forge
lz4-c                     1.9.2                h62dcd97_1    conda-forge
matplotlib                3.2.1                         0    conda-forge
matplotlib-base           3.2.1            py36hf0cddfc_0    conda-forge
mkl                       2019.4                      245
mkl-devel                 2019.4                      245
mkl-include               2019.4                      245
mkl-service               2.3.0            py36hfa6e2cd_0    conda-forge
mkl_fft                   1.1.0            py36hc8d92b1_1    conda-forge
netcdf4                   1.5.3           nompi_py36h86f0940_103    conda-forge
networkx                  2.4                        py_1    conda-forge
ninja                     1.10.0               h1ad3211_0    conda-forge
nose                      1.3.7           py36h9f0ad1d_1004    conda-forge
numexpr                   2.7.1            py36h003fed8_1    conda-forge
numpy                     1.18.5           py36h4d86e3b_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openjpeg                  2.3.1                h57dd2e7_3    conda-forge
openssl                   1.1.1g               he774522_0    conda-forge
packaging                 20.4               pyh9f0ad1d_0    conda-forge
pillow                    7.1.2            py36he4e95fe_0    conda-forge
pip                       20.1.1                     py_1    conda-forge
pycparser                 2.20                       py_0    conda-forge
pyopenssl                 19.1.0                     py_1    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
pyqt                      5.12.3           py36h003fed8_3    conda-forge
pyqt5-sip                 4.19.18                  pypi_0    pypi
pyqtchart                 5.12                     pypi_0    pypi
pyqtwebengine             5.12.1                   pypi_0    pypi
pyreadline                2.1                   py36_1001    conda-forge
pysocks                   1.7.1            py36h9f0ad1d_1    conda-forge
python                    3.6.10          he025d50_1009_cpython    conda-forge
python-coveralls          2.9.3                      py_0    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
python_abi                3.6                     1_cp36m    conda-forge
pywavelets                1.1.1            py36h7725771_1    conda-forge
pyyaml                    5.3.1            py36h68a101e_0    conda-forge
qt                        5.12.5               h7ef1ec2_0    conda-forge
requests                  2.24.0             pyh9f0ad1d_0    conda-forge
scikit-build              0.11.1           py36h003fed8_1    conda-forge
scikit-image              0.17.2           py36hcc50265_1    conda-forge
scipy                     1.3.1            py36h29ff71c_0    conda-forge
setuptools                47.3.1           py36h9f0ad1d_0    conda-forge
setuptools-scm            4.1.2              pyh9f0ad1d_0    conda-forge
setuptools_scm            4.1.2                         0    conda-forge
setuptools_scm_git_archive 1.1                        py_0    conda-forge
six                       1.15.0             pyh9f0ad1d_0    conda-forge
snappy                    1.1.8                ha925a31_2    conda-forge
spefile                   1.6                        py_1    conda-forge
sqlite                    3.30.1               hfa6e2cd_0    conda-forge
tifffile                  2020.6.3                   py_1    conda-forge
timemory                  2.3.0                    py36_0    jrmadsen
tk                        8.6.10               hfa6e2cd_0    conda-forge
toml                      0.10.1             pyh9f0ad1d_0    conda-forge
toolz                     0.10.0                     py_0    conda-forge
tornado                   6.0.4            py36hfa6e2cd_0    conda-forge
urllib3                   1.25.9                     py_0    conda-forge
vc                        14.1                 h869be7e_1    conda-forge
vs2015_runtime            14.16.27012          h30e32a0_2    conda-forge
vs2015_win-64             14.0.25420          h55c1224_11
wheel                     0.34.2                     py_1    conda-forge
win_inet_pton             1.1.0                    py36_0    conda-forge
wincertstore              0.2                   py36_1003    conda-forge
xz                        5.2.5                h2fa13f4_0    conda-forge
yaml                      0.2.5                he774522_0    conda-forge
zlib                      1.2.11            h2fa13f4_1006    conda-forge
zstd                      1.4.4                h9f78265_3    conda-forge
```

Now when I try to build it is actually using VS2015, not VS2017.  Luckily I have both installed.

However, it still does not work.  I now get this error:
```
(tomopy_devel) J:\epics\tomography\tomopy>python setup.py develop


--------------------------------------------------------------------------------
-- Trying "Visual Studio 14 2015 Win64" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- Selecting Windows SDK version  to target Windows 10.0.16299.
-- The C compiler identification is MSVC 19.0.24215.1
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe - broken
CMake Error at C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy_devel/Library/share/cmake-3.17/Modules/CMakeTestCCompiler.cmake:60 (message):
  The C compiler

    "C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: J:/epics/tomography/tomopy/_cmake_test_compile/build/CMakeFiles/CMakeTmp

    Run Build Command(s):C:/Program Files (x86)/MSBuild/14.0/bin/MSBuild.exe cmTC_4576e.vcxproj /p:Configuration=Debug /p:Platform=x64 /p:VisualStudioVersion=14.0 /v:m && Microsoft (R) Build Engine version 14.0.25420.1
    Copyright (C) Microsoft Corporation. All rights reserved.

      Microsoft (R) C/C++ Optimizing Compiler Version 19.00.24215.1 for x64
      Copyright (C) Microsoft Corporation.  All rights reserved.

      cl /c /Zi /W3 /WX- /Od /Ob0 /GL /D WIN32 /D _WINDOWS /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"cmTC_4576e.dir\Debug\\" /Fd"cmTC_4576e.dir\Debug\vc140.pdb" /Gd /TC /errorReport:queue J:\epics\tomography\tomopy\_cmake_test_compile\build\CMakeFiles\CMakeTmp\testCCompiler.c

      testCCompiler.c
    LINK : fatal error LNK1104: cannot open file 'J:\epics\tomography\tomopy\_cmake_test_compile\build\CMakeFiles\CMakeTmp\Debug\cmTC_4576e.exe' [J:\epics\tomography\tomopy\_cmake_test_compile\build\CMakeFiles\CMakeTmp\cmTC_4576e.vcxproj]





  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:3 (ENABLE_LANGUAGE)


-- Configuring incomplete, errors occurred!
See also "J:/epics/tomography/tomopy/_cmake_test_compile/build/CMakeFiles/CMakeOutput.log".
See also "J:/epics/tomography/tomopy/_cmake_test_compile/build/CMakeFiles/CMakeError.log".
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Visual Studio 14 2015 Win64" generator - failure
--------------------------------------------------------------------------------

********************************************************************************
scikit-build could not get a working generator for your system. Aborting build.

Building windows wheels for Python 3.6 requires Microsoft Visual Studio 2017.
Get it with "Visual Studio 2017":

  https://visualstudio.microsoft.com/vs/

********************************************************************************
```

The problem again seems to be that the test file is not found, there is nothing in the CMakeTmp directory.
```
(tomopy_devel) J:\epics\tomography\tomopy>dir J:\epics\tomography\tomopy\_cmake_test_compile\build\CMakeFiles\CMakeTmp\
 Volume in drive J is people_rw
 Volume Serial Number is 1FC2-CD09

 Directory of J:\epics\tomography\tomopy\_cmake_test_compile\build\CMakeFiles\CMakeTmp

06/22/2020  03:41 PM    <DIR>          .
06/22/2020  03:41 PM    <DIR>          ..
               0 File(s)              0 bytes
               2 Dir(s)  218,615,820,288 bytes free
```

I made some more progress.  

Previously I was building on my J: drive, which is a Windows mount of our Linux file system.  That normally works fine.  For example my entire EPICS development tree resides on that file system, and I build EPICS fine on it from Windows machines.  

Since I thought that might have something to do with the problem, I did a fresh clone of tomopy on my local C: drive.

Now the build gets much further before failing:
```
(tomopy_devel) C:\python\tomopy>python setup.py develop


--------------------------------------------------------------------------------
-- Trying "Visual Studio 14 2015 Win64" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- Selecting Windows SDK version  to target Windows 10.0.16299.
-- The C compiler identification is MSVC 19.0.24215.1
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/python/tomopy/_cmake_test_compile/build
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Visual Studio 14 2015 Win64" generator - success
--------------------------------------------------------------------------------

Configuring Project
  Working directory:
    C:\python\tomopy\_skbuild\win-amd64-3.6\cmake-build
  Command:
    cmake 'C:\python\tomopy' -G 'Visual Studio 14 2015 Win64' '-DCMAKE_INSTALL_PREFIX:PATH=C:\python\tomopy\_skbuild\win-amd64-3.6\cmake-install' '-DPYTHON_EXECUTABLE:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\python.exe' -DPYTHON_VERSION_STRING:STRING=3.6.10 '-DPYTHON_INCLUDE_DIR:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\Include' '-DPYTHON_LIBRARY:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\libs\python36.lib' -DSKBUILD:BOOL=TRUE '-DCMAKE_MODULE_PATH:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\lib\site-packages\skbuild\resources\cmake' -DCMAKE_BUILD_TYPE:STRING=Release

Using scikit-build. cmake: C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy_devel/Library/bin/cmake.exe...
-- Selecting Windows SDK version  to target Windows 10.0.16299.
-- The C compiler identification is MSVC 19.0.24215.1
-- The CXX compiler identification is MSVC 19.0.24215.1
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test c_W
-- Performing Test c_W - Failed
-- Performing Test c_Wextra
-- Performing Test c_Wextra - Failed
-- Performing Test c_Wno_attributes
-- Performing Test c_Wno_attributes - Failed
-- Performing Test c_Wno_unused_variable
-- Performing Test c_Wno_unused_variable - Failed
-- Performing Test c_Wno_unknown_pragmas
-- Performing Test c_Wno_unknown_pragmas - Failed
-- Performing Test c_Wno_unused_parameter
-- Performing Test c_Wno_unused_parameter - Failed
-- Performing Test c_Wno_reserved_id_macro
-- Performing Test c_Wno_reserved_id_macro - Failed
-- Performing Test c_Wunused_but_set_parameter
-- Performing Test c_Wunused_but_set_parameter - Failed
-- Performing Test c_fopenmp_simd
-- Performing Test c_fopenmp_simd - Failed
-- Performing Test cxx_fopenmp_simd
-- Performing Test cxx_fopenmp_simd - Failed
-- Performing Test c_fp_model_precise
-- Performing Test c_fp_model_precise - Failed
-- Performing Test cxx_fp_model_precise
-- Performing Test cxx_fp_model_precise - Failed
-- Performing Test cxx_W
-- Performing Test cxx_W - Failed
-- Performing Test cxx_Wextra
-- Performing Test cxx_Wextra - Failed
-- Performing Test cxx_Wno_attributes
-- Performing Test cxx_Wno_attributes - Failed
-- Performing Test cxx_Wno_unused_value
-- Performing Test cxx_Wno_unused_value - Failed
-- Performing Test cxx_Wno_unused_variable
-- Performing Test cxx_Wno_unused_variable - Failed
-- Performing Test cxx_Wno_unknown_pragmas
-- Performing Test cxx_Wno_unknown_pragmas - Failed
-- Performing Test cxx_Wno_unused_parameter
-- Performing Test cxx_Wno_unused_parameter - Failed
-- Performing Test cxx_Wno_reserved_id_macro
-- Performing Test cxx_Wno_reserved_id_macro - Failed
-- Performing Test cxx_Wno_implicit_fallthrough
-- Performing Test cxx_Wno_implicit_fallthrough - Failed
-- Performing Test cxx_Wunused_but_set_parameter
-- Performing Test cxx_Wunused_but_set_parameter - Failed
-- Performing Test cxx_faligned_new
-- Performing Test cxx_faligned_new - Failed
-- Performing Test c_MD
-- Performing Test c_MD - Success
-- Performing Test c_GL
-- Performing Test c_GL - Success
-- Performing Test cxx_MD
-- Performing Test cxx_MD - Success
-- Performing Test cxx_GL
-- Performing Test cxx_GL - Success
-- Found PythonInterp: C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy_devel/python.exe (found version "3.6.10")
-- Found MKL: C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy_devel/Library (found version "2019.0.4") found components: rt
-- OpenCV ARCH: x64
-- OpenCV RUNTIME: vc14
-- OpenCV STATIC: OFF
CMake Warning at C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy_devel/Library/cmake/OpenCVConfig.cmake:176 (message):
  Found OpenCV Windows Pack but it has no binaries compatible with your
  configuration.

  You should manually point CMake variable OpenCV_DIR to your build of OpenCV
  library.
Call Stack (most recent call first):
  cmake/Modules/Packages.cmake:65 (find_package)
  CMakeLists.txt:40 (include)


CMake Error at cmake/Modules/Packages.cmake:65 (find_package):
  Found package configuration file:

    C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy_devel/Library/cmake/OpenCVConfig.cmake

  but it set OpenCV_FOUND to FALSE so package "OpenCV" is considered to be
  NOT FOUND.
Call Stack (most recent call first):
  CMakeLists.txt:40 (include)


-- Configuring incomplete, errors occurred!
See also "C:/python/tomopy/_skbuild/win-amd64-3.6/cmake-build/CMakeFiles/CMakeOutput.log".
See also "C:/python/tomopy/_skbuild/win-amd64-3.6/cmake-build/CMakeFiles/CMakeError.log".
Traceback (most recent call last):
  File "C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\lib\site-packages\skbuild\setuptools_wrap.py", line 586, in setup
    languages=cmake_languages
  File "C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\lib\site-packages\skbuild\cmaker.py", line 240, in configure
    os.path.abspath(CMAKE_BUILD_DIR())))

An error occurred while configuring with CMake.
  Command:
    cmake 'C:\python\tomopy' -G 'Visual Studio 14 2015 Win64' '-DCMAKE_INSTALL_PREFIX:PATH=C:\python\tomopy\_skbuild\win-amd64-3.6\cmake-install' '-DPYTHON_EXECUTABLE:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\python.exe' -DPYTHON_VERSION_STRING:STRING=3.6.10 '-DPYTHON_INCLUDE_DIR:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\Include' '-DPYTHON_LIBRARY:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\libs\python36.lib' -DSKBUILD:BOOL=TRUE '-DCMAKE_MODULE_PATH:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy_devel\lib\site-packages\skbuild\resources\cmake' -DCMAKE_BUILD_TYPE:STRING=Release
  Source directory:
    C:\python\tomopy
  Working directory:
    C:\python\tomopy\_skbuild\win-amd64-3.6\cmake-build
Please see CMake's output for more information.
```
So the problem is with OpenCV.  Any idea what the problem is and how to fix it?

I tried using the win-37 environment file instead. 

I get the same error:
```
(tomopy) C:\python\tomopy>python setup.py develop


--------------------------------------------------------------------------------
-- Trying "Visual Studio 14 2015 Win64" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- Selecting Windows SDK version  to target Windows 10.0.16299.
-- The C compiler identification is MSVC 19.0.24215.1
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/python/tomopy/_cmake_test_compile/build
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Visual Studio 14 2015 Win64" generator - success
--------------------------------------------------------------------------------

Configuring Project
  Working directory:
    C:\python\tomopy\_skbuild\win-amd64-3.7\cmake-build
  Command:
    cmake 'C:\python\tomopy' -G 'Visual Studio 14 2015 Win64' '-DCMAKE_INSTALL_PREFIX:PATH=C:\python\tomopy\_skbuild\win-amd64-3.7\cmake-install' '-DPYTHON_EXECUTABLE:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\python.exe' -DPYTHON_VERSION_STRING:STRING=3.7.6 '-DPYTHON_INCLUDE_DIR:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\Include' '-DPYTHON_LIBRARY:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\libs\python37.lib' -DSKBUILD:BOOL=TRUE '-DCMAKE_MODULE_PATH:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\lib\site-packages\skbuild\resources\cmake' -DCMAKE_BUILD_TYPE:STRING=Release

Using scikit-build. cmake: C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy/Library/bin/cmake.exe...
-- Selecting Windows SDK version  to target Windows 10.0.16299.
-- The C compiler identification is MSVC 19.0.24215.1
-- The CXX compiler identification is MSVC 19.0.24215.1
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test c_W
-- Performing Test c_W - Failed
-- Performing Test c_Wextra
-- Performing Test c_Wextra - Failed
-- Performing Test c_Wno_attributes
-- Performing Test c_Wno_attributes - Failed
-- Performing Test c_Wno_unused_variable
-- Performing Test c_Wno_unused_variable - Failed
-- Performing Test c_Wno_unknown_pragmas
-- Performing Test c_Wno_unknown_pragmas - Failed
-- Performing Test c_Wno_unused_parameter
-- Performing Test c_Wno_unused_parameter - Failed
-- Performing Test c_Wno_reserved_id_macro
-- Performing Test c_Wno_reserved_id_macro - Failed
-- Performing Test c_Wunused_but_set_parameter
-- Performing Test c_Wunused_but_set_parameter - Failed
-- Performing Test c_fopenmp_simd
-- Performing Test c_fopenmp_simd - Failed
-- Performing Test cxx_fopenmp_simd
-- Performing Test cxx_fopenmp_simd - Failed
-- Performing Test c_fp_model_precise
-- Performing Test c_fp_model_precise - Failed
-- Performing Test cxx_fp_model_precise
-- Performing Test cxx_fp_model_precise - Failed
-- Performing Test cxx_W
-- Performing Test cxx_W - Failed
-- Performing Test cxx_Wextra
-- Performing Test cxx_Wextra - Failed
-- Performing Test cxx_Wno_attributes
-- Performing Test cxx_Wno_attributes - Failed
-- Performing Test cxx_Wno_unused_value
-- Performing Test cxx_Wno_unused_value - Failed
-- Performing Test cxx_Wno_unused_variable
-- Performing Test cxx_Wno_unused_variable - Failed
-- Performing Test cxx_Wno_unknown_pragmas
-- Performing Test cxx_Wno_unknown_pragmas - Failed
-- Performing Test cxx_Wno_unused_parameter
-- Performing Test cxx_Wno_unused_parameter - Failed
-- Performing Test cxx_Wno_reserved_id_macro
-- Performing Test cxx_Wno_reserved_id_macro - Failed
-- Performing Test cxx_Wno_implicit_fallthrough
-- Performing Test cxx_Wno_implicit_fallthrough - Failed
-- Performing Test cxx_Wunused_but_set_parameter
-- Performing Test cxx_Wunused_but_set_parameter - Failed
-- Performing Test cxx_faligned_new
-- Performing Test cxx_faligned_new - Failed
-- Performing Test c_MD
-- Performing Test c_MD - Success
-- Performing Test c_GL
-- Performing Test c_GL - Success
-- Performing Test cxx_MD
-- Performing Test cxx_MD - Success
-- Performing Test cxx_GL
-- Performing Test cxx_GL - Success
-- Found PythonInterp: C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy/python.exe (found version "3.7.6")
-- Found MKL: C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy/Library (found version "2019.0.4") found components: rt
-- OpenCV ARCH: x64
-- OpenCV RUNTIME: vc14
-- OpenCV STATIC: OFF
CMake Warning at C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy/Library/cmake/OpenCVConfig.cmake:176 (message):
  Found OpenCV Windows Pack but it has no binaries compatible with your
  configuration.

  You should manually point CMake variable OpenCV_DIR to your build of OpenCV
  library.
Call Stack (most recent call first):
  cmake/Modules/Packages.cmake:65 (find_package)
  CMakeLists.txt:40 (include)


CMake Error at cmake/Modules/Packages.cmake:65 (find_package):
  Found package configuration file:

    C:/Users/rivers/AppData/Local/Continuum/anaconda3/envs/tomopy/Library/cmake/OpenCVConfig.cmake

  but it set OpenCV_FOUND to FALSE so package "OpenCV" is considered to be
  NOT FOUND.
Call Stack (most recent call first):
  CMakeLists.txt:40 (include)


-- Configuring incomplete, errors occurred!
See also "C:/python/tomopy/_skbuild/win-amd64-3.7/cmake-build/CMakeFiles/CMakeOutput.log".
See also "C:/python/tomopy/_skbuild/win-amd64-3.7/cmake-build/CMakeFiles/CMakeError.log".
Traceback (most recent call last):
  File "C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\lib\site-packages\skbuild\setuptools_wrap.py", line 586, in setup
    languages=cmake_languages
  File "C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\lib\site-packages\skbuild\cmaker.py", line 240, in configure
    os.path.abspath(CMAKE_BUILD_DIR())))

An error occurred while configuring with CMake.
  Command:
    cmake 'C:\python\tomopy' -G 'Visual Studio 14 2015 Win64' '-DCMAKE_INSTALL_PREFIX:PATH=C:\python\tomopy\_skbuild\win-amd64-3.7\cmake-install' '-DPYTHON_EXECUTABLE:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\python.exe' -DPYTHON_VERSION_STRING:STRING=3.7.6 '-DPYTHON_INCLUDE_DIR:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\Include' '-DPYTHON_LIBRARY:FILEPATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\libs\python37.lib' -DSKBUILD:BOOL=TRUE '-DCMAKE_MODULE_PATH:PATH=C:\Users\rivers\AppData\Local\Continuum\anaconda3\envs\tomopy\lib\site-packages\skbuild\resources\cmake' -DCMAKE_BUILD_TYPE:STRING=Release
  Source directory:
    C:\python\tomopy
  Working directory:
    C:\python\tomopy\_skbuild\win-amd64-3.7\cmake-build
Please see CMake's output for more information.
```

--- a/INSTALL
+++ b/INSTALL
@@ -1,8 +1,0 @@
-For Windows, both MinGW-64 and `make` need to be installed and placed in the
-PATH, so that the appropriate 'gcc.exe' (that is, one that supports C99) and
-'make.exe' can be found. For Anaconda Python on Windows, adding the conda
-packages MinGW and make provides these resources.
-
-Then use
-
-    pip install .

--- a/INSTALL.rst
+++ b/INSTALL.rst
@@ -1,0 +1,25 @@
+Installing dependencies
+=======================
+To build and run TomoPy, you will need to install at least the dependencies
+listed in ``envs/{platform}-{version}.yml`` plus additional dependencies based
+on your platform. For example, installing requirements for building
+the Python 3.6 version on Linux can be accomplished as follows::
+
+    $ conda env create -f envs/linux-36.yml [-n ENVIRONMENT]
+
+
+Additional Windows Requirements
+-------------------------------
+The Windows VC++2017 compiler cannot be distributed through conda. You must
+install it using the Windows Build Tools installer. The conda package for this
+compiler merely searches for this compiler and links it to the conda
+environment into which it is installed.
+
+
+Building TomoPy
+===============
+
+After navigating to inside the `tomopy` directory, you can install TomoPy by
+running the install script in the typical Python way::
+
+    $ python setup.py install

--- a/INSTALL.rst
+++ b/INSTALL.rst
@@ -2,19 +2,27 @@ Installing dependencies
 =======================
 To build and run TomoPy, you will need to install at least the dependencies
 listed in ``envs/{platform}-{version}.yml`` plus additional dependencies based
-on your platform. For example, installing requirements for building
+on your platform. For convenience, installing requirements for building
 the Python 3.6 version on Linux can be accomplished as follows::
 
     $ conda env create -f envs/linux-36.yml [-n ENVIRONMENT]
-
+    
+This will create a new conda environment named tomopy with the build
+dependencies. If you already have a conda environment named tomopy. Use the
+`-n` option to name it something else.
 
 Additional Windows Requirements
 -------------------------------
-The Windows VC++2017 compiler cannot be distributed through conda. You must
-install it using the Windows Build Tools installer. The conda package for this
-compiler merely searches for this compiler and links it to the conda
-environment into which it is installed.
+The Windows VC++2017 compiler cannot be distributed through conda. The conda
+package for this compiler creates link from the system provided compiler
+to the conda environment into which it is installed. Install VC++2017 using
+the Windows Build Tools installer; Visual Studio (the IDE) is not required
+to build TomoPy.
 
+Additional CUDA Requirements
+----------------------------
+The CUDA compiler cannot be distributed through conda. Building TomoPy with
+GPU features requires the CUDA Toolkit and NVCC.
 
 Building TomoPy
 ===============

--- a/doc/source/devguide.rst
+++ b/doc/source/devguide.rst
@@ -34,23 +34,7 @@ and ask you where you want to save it. Select a location in your
 computer and feel comfortable with making modifications in the code.
 
 
-Installing dependencies
-=======================
-To build and run TomoPy, you will need to install at least the dependencies
-listed in ``envs/{platform}-{version}.yml`` plus additional dependencies based
-on your build environment. For example, installing requirements for building
-the Python 3.6 version on Linux can be accomplished as follows::
-
-    $ conda env create -f envs/linux-36.yml
-
-
-Building TomoPy
-===============
-
-After navigating to inside the `tomopy` directory, you can install TomoPy by
-running the install script in the typical Python way::
-
-    $ python setup.py install
+.. include:: ../../INSTALL.rst
 
 
 Running the Tests

--- a/doc/source/install.rst
+++ b/doc/source/install.rst
@@ -15,8 +15,8 @@ and its dependencies for you.
 First, you must have `Conda <https://docs.conda.io/en/latest/miniconda.html>`_
 installed.
 
-Next, install TomoPy and all its dependencies into a new Conda environment
-called ``tomopy`` by running::
+Next, install TomoPy and all its runtime dependencies into a new Conda
+environment called ``tomopy`` by running::
 
     $ conda create --name tomopy --channel conda-forge tomopy
 
@@ -36,3 +36,10 @@ containing TomoPy and run::
 
 For some more information about using Conda, please refer to the `docs
 <https://conda.io/projects/conda>`__.
+
+
+Build from Source
+-----------------
+
+Please read the development guide for directions on how to build TomoPy from
+the source code.

